### PR TITLE
refactor: set go 1.21 as minimal supported version

### PR DESCRIPTION
**Pull request by eliottness**
- [x] Upgrade go mod to 1.21
- [x] Remove go 1.20 from CI
- [x] use `errors.Join` when necessary
- [x] use builtin `max` when necessary

### CODE REVIEW

File: .github/workflows/_test_bare_metal.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
       fail-fast: false
       matrix:
         runs-on: [ macos-14, macos-13, macos-12, ubuntu-22.04, ubuntu-20.04, windows-latest, arm-4core-linux ]
-        go-version: [ '1.22', '1.21', '1.20' ]
```
Comment (RomainMuller) [inline .github/workflows/_test_bare_metal.yml:11]: No `1.23-rc` here?

File: internal/bindings/waf_dl.go [hunk 2]
```diff
@@ -43,33 +44,25 @@ type wafSymbols struct {
...
-			} else {
-				// TODO: rely on errors.Join() once go1.20 is our min supported Go version
-				err = fmt.Errorf("%w; along with an error while removing %s: %v", err, file, rmErr)
-			}
+		if rmErr := os.Remove(file); rmErr != nil {
+			err = errors.Join(err, fmt.Errorf("error removing %s: %w", file, rmErr))
 		}
 	}()
 
-	var handle uintptr
```
Comment (RomainMuller) [inline internal/bindings/waf_dl.go:63]: Why move this declaration higher up, when it's actually only useful starting there?

File: internal/lib/lib.go [hunk 3]
```diff
@@ -28,20 +29,17 @@ func DumpEmbeddedWAF() (path string, err error) {
 	path = file.Name()
 
 	defer func() {
+		allErrors := []error{err}
```
Comment (RomainMuller) [inline internal/lib/lib.go:32]: IIRC, you could just `err = errors.Join(err, newErr)` and spare the slice allocation?

**Review by RomainMuller — APPROVED** [review]
Nitpicks are take it or leave it :)

File: .github/workflows/_test_bare_metal.yml (hunk 1 above)
Comment (eliottness) [reply .github/workflows/_test_bare_metal.yml:11]: Nope, cf. [this](https://github.com/actions/go-versions/blob/main/versions-manifest.json)

File: internal/bindings/waf_dl.go (hunk 2 above)
Comment (eliottness) [reply internal/bindings/waf_dl.go]: Done

File: internal/lib/lib.go (hunk 3 above)
Comment (eliottness) [reply internal/lib/lib.go]: Done

**Pull request merged**